pilot/model/validation_test.go

```diff
@@ -1303,23 +1335,60 @@ var (
 func TestValidateMixerAttributes(t *testing.T) {
 	cases := []struct {
 		name  string
-		in    proto.Message
+		in    *mpb.Attributes_AttributeValue
 		valid bool
 	}{
-		{
-			name:  "valid",
-			in:    validAttributes,
-			valid: true,
-		},
-		{
-			name: "invalid",
-			in:   invalidAttributes,
-		},
+		{"happy string",
+			&mpb.Attributes_AttributeValue{Value: &mpb.Attributes_AttributeValue_StringValue{"my-service"}},
+			true},
+		{"invalid string",
+			&mpb.Attributes_AttributeValue{Value: &mpb.Attributes_AttributeValue_StringValue{""}},
+			false},
+		{"happy duration",
+			&mpb.Attributes_AttributeValue{Value: &mpb.Attributes_AttributeValue_DurationValue{&types.Duration{Nanos: -1}}},
+			true},
+		{"invalid duration",
+			&mpb.Attributes_AttributeValue{Value: &mpb.Attributes_AttributeValue_DurationValue{&types.Duration{Nanos: -1e9}}},
+			false},
+		{"nil duration",
+			&mpb.Attributes_AttributeValue{Value: &mpb.Attributes_AttributeValue_DurationValue{nil}},
+			false},
+		{"happy bytes",
+			&mpb.Attributes_AttributeValue{Value: &mpb.Attributes_AttributeValue_BytesValue{[]byte{1, 2, 3}}},
+			true},
+		{"invalid bytes",
+			&mpb.Attributes_AttributeValue{Value: &mpb.Attributes_AttributeValue_BytesValue{[]byte{}}},
+			false},
+		{"happy timestamp",
+			&mpb.Attributes_AttributeValue{Value: &mpb.Attributes_AttributeValue_TimestampValue{&types.Timestamp{}}},
+			true},
+		{"invalid timestamp",
+			&mpb.Attributes_AttributeValue{Value: &mpb.Attributes_AttributeValue_TimestampValue{&types.Timestamp{Nanos: -1}}},
+			false},
+		{"nil timestamp",
+			&mpb.Attributes_AttributeValue{Value: &mpb.Attributes_AttributeValue_TimestampValue{nil}},
+			false},
+		{"happy stringmap",
+			&mpb.Attributes_AttributeValue{Value: &mpb.Attributes_AttributeValue_StringMapValue{
+				&mpb.Attributes_StringMap{Entries: map[string]string{"foo": "bar"}}}},
+			true},
+		{"invalid stringmap",
+			&mpb.Attributes_AttributeValue{Value: &mpb.Attributes_AttributeValue_StringMapValue{
+				&mpb.Attributes_StringMap{Entries: nil}}},
+			false},
+		{"nil stringmap",
+			&mpb.Attributes_AttributeValue{Value: &mpb.Attributes_AttributeValue_StringMapValue{nil}},
+			false},
 	}
 	for _, c := range cases {
-		if got := ValidateMixerAttributes(c.in); (got == nil) != c.valid {
-			t.Errorf("ValidateMixerAttributes(%v): got(%v) != want(%v): %v", c.name, got == nil, c.valid, got)
-		}
+		t.Run(c.name, func(t *testing.T) {
+			attrs := &mpb.Attributes{
+				Attributes: map[string]*mpb.Attributes_AttributeValue{"key": c.in},
+			}
+			if got := ValidateMixerAttributes(attrs); (got == nil) != c.valid {
+				t.Errorf("ValidateMixerAttributes(%v): got(%v) != want(%v): %v", c.name, got == nil, c.valid, got)
```

Comment:
Is `ValidateMixerAttribute(%v)` and the explicit name needed here with subtests? I believe omitting these may give more concise test output, e.g.

```bash
--- FAIL: TestValidateMixerAttributes (0.00s)
    --- PASS: TestValidateMixerAttributes/happy duration (0.00s)
    --- FAIL: TestValidateMixerAttributes/nil duration (0.00s) 
    got(...) != want(...)   
```